Optimize compilation events to prevent unneeded stats.toJson() call

MLoughry opened the pull request:
<!--
  ZOMG a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and @ken_wheeler
  will appear and pile-drive the close button from a great height while making
  animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->

This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [ ] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->

Fixes #181 to prevent unneeded calls to `stats.toJson()`

### Breaking Changes

None
<!--
  If this PR introduces a breaking change, please describe the impact and a
  migration path for existing applications.
-->

### Additional Info

None

shellscape commented:
@MLoughry please don't remove portions of issue/pull request templates. we'll review once your pull request template is complete, and your changes are updated to pass CI.

MLoughry commented:
I updated to fix tests by changing away from using `nanobus` to using Node's native `EventEmitter`.

Per nanobus's documentation:
> ### When shouldn't I use this package?
> If you're only writing code that runs inside Node and don't need a '*' listener, consider using the built-in event emitter API instead.

I also added the parts of the bug template back that I had deleted, even though I deleted them precisely because there was nothing to say in those sections.

shellscape commented:
Thanks for adding the template back, and for the PR 🍺 

> I also added the parts of the bug template back that I had deleted, even though I deleted them precisely because there was nothing to say in those sections.

They aren't there to pick and choose. `None` or `n/a` is acceptable. You'll find that for most projects that provide templates, removing portions of them is frowned upon. Templates are there to make the lives of maintainers and passers-by a little bit easier.

> I updated to fix tests by changing away from using `nanobus` to using Node's native `EventEmitter`.

That's a major bummer. We won't be able to approve this PR with that change.

shellscape commented on lib/bus.js:
```diff
@@ -1,12 +1,12 @@
 const isPlainObject = require('lodash/isPlainObject');
-const nanobus = require('nanobus');
+const EventEmitter = require('events');
```
please revert this back to `nanobus`

MLoughry replied:
It would be good to include comments as to why `nanobus` is being used, when even their documentation would indicate that this project should use the native `EventEmitter`.

Regardless, `nanobus` luckily does have a `listeners()` API, so I'm able to work around the issue.

shellscape commented on index.js:
```diff
@@ -38,13 +38,21 @@ module.exports = (opts) => {
-
-          if (stats.hasWarnings()) {
-            bus.emit('compiler-warning', { json, compiler: comp });
+          const compilerErrorEventName = 'compiler-error';
```
this is very microsofty :)

would prefer:

```
const events = {
  error: 'compiler-error',
  warning: 'compiler-warning',
};
```

MLoughry replied:
Done

MLoughry commented:
I've addressed both comments from the previous review.

shellscape commented:
@MLoughry looks good. I made some slight alterations with a commit. Please let me know how you feel about it. _(I'm bouncing between contexts like a madman today and it was quicker to sprinkle my sugar on it than ask you for another change)_

MLoughry commented:
It looks fine.
